feat(CategoryTheory): Basics of Locally Cartesian Closed Categories

sinhp opened the pull request:
Supersedes https://github.com/leanprover-community/mathlib4/pull/22321

This PR defines locally cartesian closed categories in terms of existence of the pushforward functors (right adjoint to the pullback functor) for all morphisms. We develop basic API and prove the following:

1.  Existence of the pushforward functors is equivalent to cartesian closed slices. 
2. Any locally cartesian closed category with a terminal object is cartesian closed.
3. The slices of a locally cartesian closed category are locally cartesian closed.

---
- [ ] depends on: #31033 
- [ ] depends on: #30373
- [ ] depends on: #31332 

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

sinhp committed:
the preliminaries for locally cartesian closed categories (Sigma, Reindex, pull-push)

(Sigma=push, Reindex=pull, pull-push = cartesian product)

sinhp committed:
basics of locally cartesian closed categories

doc string fix

fix

basics of lcccs

Create ExponentiableMorphism.lean

Update Mathlib.lean

Update ExponentiableMorphism.lean

Update Mathlib/CategoryTheory/LocallyCartesianClosed/ExponentiableMorphism.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

rm exp-mor stuff

reference

ofOverExponentiable

cartesian monoidal from ChosenPullback

sinhp committed:
The section functor as a right adjoint to the star functor

sinhp committed:
The section functor as a right adjoint to the star functor

sinhp committed:
The section functor as a right adjoint to the star functor

sinhp committed:
fix

sinhp committed:
Merge remote-tracking branch 'refs/remotes/sina/lccc-prelim' into lccc-prelim

sinhp committed:
prelims for lcccs

sinhp committed:
fix for HasPullbacksAlong

sinhp committed:
parentheses and simp

sinhp committed:
Update Mathlib.lean

sinhp committed:
not needed assumption

sinhp committed:
rm simp

sinhp committed:
correcting attribution

sinhp committed:
Update Prelim.lean

sinhp committed:
Merge branch 'lccc-section' of https://github.com/sinhp/mathlib4 into lccc-section

Jlh18 commented:
I am currently working on generalizing the definition of pushforward. See PR #30754. It doesn't require all pushforwards to exist, i.e. you may only have a partial right adjoint of pullback

sinhp committed:
ChosenPullback

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
pullbackObj

sinhp committed:
pullbackObj

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
ChosenPullback fst

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
ChosenPullback.hasPullbackAlong

sinhp committed:
ChosenPullback.pullback iso Over.pullback

sinhp committed:
remove second have

sinhp committed:
rm simp

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
rm have

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
naming

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
terminal simp

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
rm simp!

sinhp committed:
ChosenPullbacks C

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
HasPullbacks instance from ChosenPullbacks

sinhp committed:
Update ChosenPullback.lean

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
remove overMkHom

sinhp committed:
doc string

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
doc string added

sinhp committed:
Create types.lean

sinhp committed:
rm types from this

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
Create Types.lean

sinhp committed:
Delete Mathlib/CategoryTheory/LocallyCartesianClosed/Types.lean

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
pullbackIsoOverPullback_inv_app_

sinhp committed:
generalize pullbackMap

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
simp nf

sinhp committed:
Merge branch 'lccc-section' of https://github.com/sinhp/mathlib4 into lccc-section

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into lccc-chosen_pullback

sinhp committed:
rm mapPullbackDesc

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
proof wo convert

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
doc string added

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
namespaces

sinhp committed:
Merge branch 'lccc-chosen_pullback' into lccc-section

sinhp committed:
computable star

sinhp committed:
Merge branch 'leanprover-community:master' into lccc-section

sinhp committed:
Delete Mathlib/CategoryTheory/LocallyCartesianClosed/Prelim.lean

sinhp committed:
Delete Mathlib/CategoryTheory/LocallyCartesianClosed/Sections0.lean

sinhp committed:
Update Sections.lean

sinhp committed:
Merge branch 'lccc-section' of https://github.com/sinhp/mathlib4 into lccc-section

sinhp committed:
Update Sections.lean

sinhp committed:
computable sections

sinhp committed:
use (Over.forget ..).mapIso

sinhp committed:
Merge branch 'leanprover-community:master' into lccc-section

sinhp committed:
toOver

sinhp committed:
linter

sinhp committed:
fix the damn linter; lake exe mk_all to get the order of names "right"

yikes

sinhp committed:
rm junk

sinhp committed:
ext instead of hom_ext

sinhp committed:
change fst to snd

this is because, as seen in lccc-section, we want to have
ChosenPullback.snd (toUnit X) (toUnit I) = CartesianMonoidalCategory.snd X I := rfl

sinhp committed:
doc string

sinhp committed:
Squashed version of lccc-chosen_pullback

sinhp committed:
Merge branch 'lccc-chosen_pullback-squashed' into lccc-section

sinhp committed:
not needed

sinhp committed:
Merge branch 'master' into lccc-section

sinhp committed:
Merge remote-tracking branch 'origin/master' into lccc-section

sinhp committed:
naming

sinhp committed:
naming

sinhp committed:
Merge branch 'lccc-chosen_pullback' into lccc-section

sinhp committed:
lemmas on interaction of toOver and ChosenPullback.pullback

sinhp committed:
toOverTerminal from toOverUnit

sinhp committed:
Create ExponentiableMorphism.lean

sinhp committed:
computable cartesian monoidal for slices

sinhp committed:
pullbackId and pullbackComp

sinhp committed:
rm junk

sinhp committed:
fixup! rm junk

sinhp committed:
Merge branch 'lccc-chosen_pullback' into lccc-section

sinhp committed:
make it computable

sinhp committed:
isoInv

sinhp committed:
Merge branch 'lccc-chosen_pullback' into lccc-section

sinhp committed:
variable scoping

sinhp committed:
add docstring

sinhp committed:
Update ChosenPullback.lean

sinhp committed:
Merge branch 'lccc-chosen_pullback' into lccc-section

sinhp committed:
rm unused import

sinhp committed:
cartesian monoidal section moved to a new PR

sinhp committed:
remove namespace Over

sinhp committed:
computable cartesian monoidal structure on slices

sinhp committed:
exe mk_all

sinhp committed:
rename to ChosenPullbackAlong

sinhp committed:
improve docstrings

sinhp committed:
Delete Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

sinhp committed:
add s

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
s

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
name change

sinhp committed:
Update Over.lean

sinhp committed:
Merge branch 'master' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into lccc-over

sinhp committed:
sorry free

sinhp committed:
equivToOverUnit

sinhp committed:
Update Over.lean

sinhp committed:
remove simp from equality of functors

https://github.com/leanprover-community/mathlib4/pull/30373#discussion_r2526336070

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
binaryFan_fst & binaryFan_snd

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
explicit vars

sinhp committed:
Merge branch 'lccc-over' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
Merge branch 'master' of https://github.com/sinhp/mathlib4 into lccc-section

sinhp committed:
remove old merged file

sinhp committed:
Merge branch 'lccc-over' into lccc-section

sinhp committed:
fix after merge

sinhp committed:
abbrev

sinhp committed:
BraidedCategory C

sinhp committed:
Merge branch 'master' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
rm mixin lemmas

sinhp committed:
fix names

sinhp committed:
Update Mathlib.lean

sinhp committed:
namespace cleanup

sinhp committed:
Merge branch 'lccc-exp_mor' into lccc-basic

dagurtomas committed:
Merge branch 'master' into lccc-over

dagurtomas committed:
Apply suggestion from @dagurtomas

dagurtomas committed:
move `@[expose] public section`

Expose the section for better accessibility.

sinhp committed:
Merge branch 'master' into lccc-over

sinhp committed:
make everything computable

sinhp committed:
Merge branch 'master' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
Merge branch 'lccc-over' into lccc-exp_mor

sinhp committed:
new module system

sinhp committed:
don't need Over

sinhp committed:
not needed in this PR

sinhp committed:
Update ExponentiableMorphism.lean

sinhp committed:
Merge branch 'lccc-over' into lccc-section

sinhp committed:
Update Sections.lean

sinhp committed:
Update Mathlib.lean

sinhp committed:
Update Mathlib.lean

sinhp committed:
Update Mathlib.lean

sinhp committed:
simp nf

sinhp committed:
Update ExponentiableMorphism.lean

sinhp committed:
@[expose] public section

sinhp committed:
Merge branch 'lccc-over' into lccc-basic

sinhp committed:
imports

sinhp committed:
Merge branch 'lccc-section' into lccc-basic

sinhp committed:
Merge branch 'lccc-exp_mor' into lccc-basic

sinhp committed:
Create Over.lean

sinhp committed:
move isLimitPullbackCone

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Over.lean

sinhp committed:
not needed

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Dagur Asgeirsson <dagurtomas@gmail.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Dagur Asgeirsson <dagurtomas@gmail.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Dagur Asgeirsson <dagurtomas@gmail.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Dagur Asgeirsson <dagurtomas@gmail.com>

sinhp committed:
remove simpa using

sinhp committed:
Merge branch 'master' into lccc-over

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullbacksAlong.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Merge branch 'master' into lccc-over

sinhp committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into lccc-section

sinhp committed:
Merge branch 'master' into lccc-exp_mor

sinhp committed:
@[expose] public section

sinhp committed:
Delete Over.lean

sinhp committed:
Merge branch 'master' into lccc-exp_mor

sinhp committed:
fix

sinhp committed:
Merge branch 'master' of https://github.com/sinhp/mathlib4 into lccc-basic

sinhp committed:
Merge branch 'lccc-over' into lccc-basic

sinhp committed:
Merge branch 'lccc-exp_mor' into lccc-basic

sinhp committed:
Merge branch 'lccc-section' into lccc-basic

sinhp committed:
refactor

sinhp committed:
Update Basic.lean

sinhp committed:
Update Basic.lean

sinhp committed:
fix linter error
